### PR TITLE
SHA-256 and SHA-512 support

### PR DESCRIPTION
**Pull request by qqmyers**
## New Contributors

Welcome! New contributors should at least glance at [CONTRIBUTING.md](/CONTRIBUTING.md), especially the section on pull requests where we encourage you to reach out to other developers before you start coding. Also, please note that we measure code coverage and prefer you write unit tests. Pull requests can still be reviewed without tests or completion of the checklist outlined below. Thanks!

## Related Issues

- connects to #5033 - Support for SHA-256 and SHA-512

## Pull Request Checklist

- [ ] Unit [tests][] completed
- [ ] Integration [tests][]: None
- [ ] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [ ] [Documentation][docs] completed
- [ ] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: https://github.com/IQSS/dataverse/tree/develop/scripts/database/upgrades
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

File: doc/sphinx-guides/source/installation/config.rst [hunk 1]
```diff
@@ -1089,7 +1089,11 @@ This sets the base name (without dot and extension), if not set it defaults to '
 
 Dataverse calculates checksums for uploaded files so that users can determine if their file was corrupted via upload or download. This is sometimes called "file fixity": https://en.wikipedia.org/wiki/File_Fixity
 
-The default checksum algorithm used is MD5 and should be sufficient for establishing file fixity. "SHA-1" is an experimental alternate value for this setting.
+The default checksum algorithm used is MD5 and should be sufficient for establishing file fixity. "SHA-1", "SHA-256" and "SHA-512" are alternate values for this setting.
+The fixity algorithm used on existing files can be changed by a superuser using the API. The API call will only update the algorithm and checksum for a file if the existing checksum can be validated against the file.
+Statistics concerning the updates are returned in the response to the API call with details in the log.
+
+``curl -X PUT -d domainstats http://localhost:8080/api/admin/updateHashValues/{alg}``
```
Comment (pdurbin) [inline doc/sphinx-guides/source/installation/config.rst]: Is "domainstats" a typo?

Comment (pdurbin) [reply doc/sphinx-guides/source/installation/config.rst]: This config section may not be the best place to document "admin" commands. I'm wondering if it should go in a future version of one of these places:

- http://guides.dataverse.org/en/4.9.2/admin/dataverses-datasets.html
- http://guides.dataverse.org/en/4.9.2/api/native-api.html#admin

Comment (pdurbin) [reply doc/sphinx-guides/source/installation/config.rst]: While we're in here, we should add a `curl` example for `:FileFixityChecksumAlgorithm`.

File: src/main/java/edu/harvard/iq/dataverse/api/Admin.java [hunk 2]
```diff
@@ -1099,6 +1104,91 @@ public Response registerDataFileAll() {
...
+		} catch (IllegalArgumentException iae) {
+			return error(Status.BAD_REQUEST, "Unknown algorithm");
+		}
+		logger.info("Starting to rehash: analyzing " + count + " files. " + new Date());
+		logger.info("Hashes not created with " + alg + " will be verified, and, if valid, replaced with a hash using "
+				+ alg);
+		try {
+			User u = findAuthenticatedUserOrDie();
+			if (!u.isSuperuser())
+				return error(Status.UNAUTHORIZED, "must be superuser");
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/api/Admin.java]: I'm glad to see this is superuser-only for now.

File: src/main/java/edu/harvard/iq/dataverse/api/Admin.java [hunk 3]
```diff
@@ -1099,6 +1104,91 @@ public Response registerDataFileAll() {
...
+				logger.info("Unexpected Exception: " + e.getMessage());
+
+			}
+		}
+		logger.info("Final Results:");
+		logger.info(
+				alreadyUpdated + " of  " + count + " files already had hashes with the new algorithm. " + new Date());
+		logger.info(rehashed + " of  " + count + " files to rehash. " + new Date());
+		logger.info(
+				successes + " of  " + rehashed + " files successfully rehashed with the new algorithm. " + new Date());
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/api/Admin.java]: This is a lot of logging at the "info" level. Should it be reduced to "fine"?

**Comment by qqmyers** [comment]
@pdurbin - I made some updates to address your comments. In general, feel free to change documentation or logging (or anything) - I'll watch for issues but you guys should have edit access and if its faster to make changes than to iterate with me, just go for it.

**Comment by pdurbin** [comment]
@qqmyers how do you feel about me reformatting the code with Netbeans?

**Comment by qqmyers** [comment]
@pdurbin - should have added formatting in the list above - feel free. (For the Admin file, I was going to swap tab for 4 spaces and I'm trying to stick with spaces overall now, but it looked like the rest of the file used tabs, outside my changes on this branch (possibly me from some earlier merge).)

File: src/main/java/edu/harvard/iq/dataverse/api/Admin.java (hunk 3 above)
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/api/Admin.java]: Actually, I was going to suggest that we keep the more verbose logging. This is a very special operation; nobody's going to be changing the checksum algorithms on their files every day. And it's something important enough, so could be a good thing to have some "paper trail" of this switch having been performed.

Comment (qqmyers) [reply src/main/java/edu/harvard/iq/dataverse/api/Admin.java]: Up to you guys. As is now, the summary is info level and there are warnings for any file that doesn't match (or can't be read, etc.). The original hash is deleted in this operation though, so if keeping the original one documented is trail-worthy, a per-file info message could be added with the original algorithm and hash and file id (storageidentifier?).

**Pull request merged**